File: FRAMS/lib/faceEncodingConverter.ts
/**
 * Face Encoding Converter Service
 * Handles bidirectional conversion between face_recognition (128-dim) and MediaPipe (1404-dim) formats
 */

export type EncodingFormat = 'face_recognition' | 'mediapipe';
export type FaceEncoding = number[];

export interface EncodingMetadata {
  encoding: FaceEncoding;
  format: EncodingFormat;
  source_system: 'FRAMS' | 'Face_Reco';
  created_at: string;
  conversion_quality?: number;
  original_dimensions?: number;
}

export interface ConversionResult {
  success: boolean;
  encoding?: EncodingMetadata;
  quality_score?: number;
  error?: string;
}

/**
 * Face Encoding Converter Class
 * Provides methods to convert between different face encoding formats
 */
export class FaceEncodingConverter {
  private static readonly FACE_RECOGNITION_DIM = 128;
  private static readonly MEDIAPIPE_DIM = 1404; // 468 landmarks × 3 coordinates (x,y,z)

  /**
   * Convert MediaPipe encoding (1404-dim) to face_recognition format (128-dim)
   * @param mediapipeEncoding - 1404-dimensional MediaPipe face mesh encoding
   * @returns Converted 128-dimensional encoding
   */
  static mediapipeToFaceRecognition(mediapipeEncoding: FaceEncoding): ConversionResult {
    try {
      // Validate input
      if (!Array.isArray(mediapipeEncoding)) {
        return {
          success: false,
          error: 'Input must be an array'
        };
      }

      if (mediapipeEncoding.length !== this.MEDIAPIPE_DIM) {
        return {
          success: false,
          error: `Expected ${this.MEDIAPIPE_DIM}-dimensional encoding, got ${mediapipeEncoding.length}`
        };
      }

      // Extract key facial landmarks from MediaPipe data
      // MediaPipe provides 468 landmarks with x,y,z coordinates
      // We'll extract the most discriminative features
      
      const convertedEncoding: number[] = [];
      
      // Strategy: Extract statistical features from landmark groups
      // Group landmarks by facial regions and compute statistical measures
      
      // 1. Extract nose bridge landmarks (landmarks 1-10)
      const noseLandmarks = mediapipeEncoding.slice(0, 30); // First 10 landmarks × 3 coords
      convertedEncoding.push(...this.computeStatisticalFeatures(noseLandmarks, 10));
      
      // 2. Extract eye region landmarks (landmarks 33-158 for left eye, 362-487 for right eye)
      const leftEyeLandmarks = mediapipeEncoding.slice(99, 474); // Landmarks 33-158
      const rightEyeLandmarks = mediapipeEncoding.slice(1086, 1461); // Landmarks 362-487
      convertedEncoding.push(...this.computeStatisticalFeatures(leftEyeLandmarks, 15));
      convertedEncoding.push(...this.computeStatisticalFeatures(rightEyeLandmarks, 15));
      
      // 3. Extract mouth region landmarks (landmarks 61-78, 308-324)
      const mouthLandmarks = [
        ...mediapipeEncoding.slice(183, 234), // Landmarks 61-78
        ...mediapipeEncoding.slice(924, 975)  // Landmarks 308-324
      ];
      convertedEncoding.push(...this.computeStatisticalFeatures(mouthLandmarks, 12));
      
      // 4. Extract jaw/chin landmarks (landmarks 14-17, 199-200)
      const jawLandmarks = [
        ...mediapipeEncoding.slice(42, 51),   // Landmarks 14-17
        ...mediapipeEncoding.slice(597, 600)  // Landmarks 199-200
      ];
      convertedEncoding.push(...this.computeStatisticalFeatures(jawLandmarks, 8));
      
      // 5. Extract cheek landmarks (landmarks 123-143, 352-372)
      const cheekLandmarks = [
        ...mediapipeEncoding.slice(369, 429), // Landmarks 123-143
        ...mediapipeEncoding.slice(1056, 1116) // Landmarks 352-372
      ];
      convertedEncoding.push(...this.computeStatisticalFeatures(cheekLandmarks, 15));
      
      // 6. Global facial shape features
      convertedEncoding.push(...this.computeGlobalFeatures(mediapipeEncoding));
      
      // Pad or truncate to exactly 128 dimensions
      while (convertedEncoding.length < this.FACE_RECOGNITION_DIM) {
        convertedEncoding.push(0);
      }
      
      if (convertedEncoding.length > this.FACE_RECOGNITION_DIM) {
        convertedEncoding.splice(this.FACE_RECOGNITION_DIM);
      }
      
      // Normalize the encoding
      const normalizedEncoding = this.normalizeEncoding(convertedEncoding);
      
      // Calculate quality score based on conversion fidelity
      const qualityScore = this.calculateConversionQuality(
        mediapipeEncoding, 
        normalizedEncoding
      );
      
      return {
        success: true,
        encoding: {
          encoding: normalizedEncoding,
          format: 'face_recognition',
          source_system: 'Face_Reco',
          created_at: new Date().toISOString(),
          conversion_quality: qualityScore,
          original_dimensions: mediapipeEncoding.length
        },
        quality_score: qualityScore
      };
      
    } catch (error) {
      return {
        success: false,
        error: `Conversion failed: ${(error as Error).message}`
      };
    }
  }

  /**
   * Convert face_recognition encoding (128-dim) to MediaPipe format (1404-dim)
   * Note: This is a lossy conversion with lower accuracy
   * @param faceRecognitionEncoding - 128-dimensional face_recognition encoding
   * @returns Converted 1404-dimensional encoding
   */
  static faceRecognitionToMediapipe(faceRecognitionEncoding: FaceEncoding): ConversionResult {
    try {
      // Validate input
      if (!Array.isArray(faceRecognitionEncoding)) {
        return {
          success: false,
          error: 'Input must be an array'
        };
      }

      if (faceRecognitionEncoding.length !== this.FACE_RECOGNITION_DIM) {
        return {
          success: false,
          error: `Expected ${this.FACE_RECOGNITION_DIM}-dimensional encoding, got ${faceRecognitionEncoding.length}`
        };
      }

      // This is a reconstruction attempt - we'll distribute the 128 features
      // across the 1404-dimensional space using interpolation and duplication
      const mediapipeEncoding: number[] = new Array(this.MEDIAPIPE_DIM).fill(0);
      
      // Distribute the 128 features across 468 landmark positions
      const featuresPerLandmark = Math.floor(this.FACE_RECOGNITION_DIM / 468);
      const remainingFeatures = this.FACE_RECOGNITION_DIM % 468;
      
      let featureIndex = 0;
      
      // Assign features to landmarks
      for (let landmark = 0; landmark < 468; landmark++) {
        const startIndex = landmark * 3;
        const endIndex = Math.min(startIndex + 3, this.MEDIAPIPE_DIM);
        
        // Assign available features to x,y,z coordinates
        for (let coord = startIndex; coord < endIndex && featureIndex < this.FACE_RECOGNITION_DIM; coord++) {
          mediapipeEncoding[coord] = faceRecognitionEncoding[featureIndex];
          featureIndex++;
        }
      }
      
      // Distribute remaining features
      for (let i = 0; i < remainingFeatures && featureIndex < this.FACE_RECOGNITION_DIM; i++) {
        const position = Math.floor((i / remainingFeatures) * this.MEDIAPIPE_DIM);
        mediapipeEncoding[position] = faceRecognitionEncoding[featureIndex];
        featureIndex++;
      }
      
      // Apply smoothing to make the encoding more realistic
      mediapipeEncoding.forEach((_, index) => {
        if (index > 2 && index < mediapipeEncoding.length - 3) {
          mediapipeEncoding[index] = (
            mediapipeEncoding[index - 3] +
            mediapipeEncoding[index - 2] +
            mediapipeEncoding[index - 1] +
            mediapipeEncoding[index] +
            mediapipeEncoding[index + 1] +
            mediapipeEncoding[index + 2] +
            mediapipeEncoding[index + 3]
          ) / 7;
        }
      });
      
      // Normalize to valid coordinate ranges (0-1 for MediaPipe)
      const normalizedEncoding = this.normalizeToCoordinateRange(mediapipeEncoding);
      
      return {
        success: true,
        encoding: {
          encoding: normalizedEncoding,
          format: 'mediapipe',
          source_system: 'FRAMS',
          created_at: new Date().toISOString(),
          conversion_quality: 0.65, // Lower quality due to lossy conversion
          original_dimensions: faceRecognitionEncoding.length
        },
        quality_score: 0.65
      };
      
    } catch (error) {
      return {
        success: false,
        error: `Conversion failed: ${(error as Error).message}`
      };
    }
  }

  /**
   * Validate face encoding format and integrity
   * @param encoding - Face encoding to validate
   * @param format - Expected format
   * @returns Validation result
   */
  static validateEncoding(encoding: FaceEncoding, format: EncodingFormat): { valid: boolean; error?: string } {
    if (!Array.isArray(encoding)) {
      return { valid: false, error: 'Encoding must be an array' };
    }

    if (format === 'face_recognition' && encoding.length !== this.FACE_RECOGNITION_DIM) {
      return { 
        valid: false, 
        error: `Face recognition encoding must be ${this.FACE_RECOGNITION_DIM}-dimensional` 
      };
    }

    if (format === 'mediapipe' && encoding.length !== this.MEDIAPIPE_DIM) {
      return { 
        valid: false, 
        error: `MediaPipe encoding must be ${this.MEDIAPIPE_DIM}-dimensional` 
      };
    }

    // Check for NaN or infinite values
    for (const value of encoding) {
      if (!isFinite(value) || isNaN(value)) {
        return { valid: false, error: 'Encoding contains invalid numeric values' };
      }
    }

    return { valid: true };
  }

  /**
   * Compute statistical features from landmark data
   * @param landmarks - Array of landmark coordinates
   * @param numFeatures - Number of features to extract
   * @returns Statistical features
   */
  private static computeStatisticalFeatures(landmarks: number[], numFeatures: number): number[] {
    if (landmarks.length === 0) return new Array(numFeatures).fill(0);
    
    const features: number[] = [];
    
    // Mean
    const mean = landmarks.reduce((sum, val) => sum + val, 0) / landmarks.length;
    features.push(mean);
    
    // Standard deviation
    const variance = landmarks.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / landmarks.length;
    features.push(Math.sqrt(variance));
    
    // Min and Max
    features.push(Math.min(...landmarks));
    features.push(Math.max(...landmarks));
    
    // Range
    features.push(Math.max(...landmarks) - Math.min(...landmarks));
    
    // Additional statistical moments
    const skewness = this.computeSkewness(landmarks, mean);
    features.push(skewness);
    
    const kurtosis = this.computeKurtosis(landmarks, mean, Math.sqrt(variance));
    features.push(kurtosis);
    
    // Percentile features
    const sorted = [...landmarks].sort((a, b) => a - b);
    features.push(sorted[Math.floor(sorted.length * 0.25)]); // 25th percentile
    features.push(sorted[Math.floor(sorted.length * 0.75)]); // 75th percentile
    
    // Pad or truncate to required number of features
    while (features.length < numFeatures) {
      features.push(0);
    }
    
    return features.slice(0, numFeatures);
  }

  /**
   * Compute global facial features
   * @param mediapipeEncoding - Full MediaPipe encoding
   * @returns Global features
   */
  private static computeGlobalFeatures(mediapipeEncoding: number[]): number[] {
    const features: number[] = [];
    
    // Overall facial width (distance between leftmost and rightmost points)
    let minX = Infinity, maxX = -Infinity;
    for (let i = 0; i < mediapipeEncoding.length; i += 3) {
      if (mediapipeEncoding[i] < minX) minX = mediapipeEncoding[i];
      if (mediapipeEncoding[i] > maxX) maxX = mediapipeEncoding[i];
    }
    features.push(maxX - minX);
    
    // Overall facial height (distance between topmost and bottommost points)
    let minY = Infinity, maxY = -Infinity;
    for (let i = 1; i < mediapipeEncoding.length; i += 3) {
      if (mediapipeEncoding[i] < minY) minY = mediapipeEncoding[i];
      if (mediapipeEncoding[i] > maxY) maxY = mediapipeEncoding[i];
    }
    features.push(maxY - minY);
    
    // Facial depth variation
    let minZ = Infinity, maxZ = -Infinity;
    for (let i = 2; i < mediapipeEncoding.length; i += 3) {
      if (mediapipeEncoding[i] < minZ) minZ = mediapipeEncoding[i];
      if (mediapipeEncoding[i] > maxZ) maxZ = mediapipeEncoding[i];
    }
    features.push(maxZ - minZ);
    
    return features;
  }

  /**
   * Normalize encoding to standard range
   * @param encoding - Encoding to normalize
   * @returns Normalized encoding
   */
  private static normalizeEncoding(encoding: number[]): number[] {
    if (encoding.length === 0) return encoding;
    
    const mean = encoding.reduce((sum, val) => sum + val, 0) / encoding.length;
    const stdDev = Math.sqrt(
      encoding.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / encoding.length
    );
    
    // Avoid division by zero
    if (stdDev === 0) return encoding.map(() => 0);
    
    return encoding.map(val => (val - mean) / stdDev);
  }

  /**
   * Normalize to MediaPipe coordinate range (0-1)
   * @param encoding - Encoding to normalize
   * @returns Normalized encoding
   */
  private static normalizeToCoordinateRange(encoding: number[]): number[] {
    if (encoding.length === 0) return encoding;
    
    const minVal = Math.min(...encoding);
    const maxVal = Math.max(...encoding);
    const range = maxVal - minVal;
    
    if (range === 0) return encoding.map(() => 0.5);
    
    return encoding.map(val => (val - minVal) / range);
  }

  /**
   * Calculate conversion quality score
   * @param original - Original encoding
   * @param converted - Converted encoding
   * @returns Quality score (0-1)
   */
  private static calculateConversionQuality(original: number[], converted: number[]): number {
    // For MediaPipe to face_recognition conversion, we can't perfectly measure quality
    // since it's a dimensionality reduction, but we can estimate based on:
    // 1. Preservation of statistical properties
    // 2. Consistency of the conversion
    
    const origMean = original.reduce((sum, val) => sum + val, 0) / original.length;
    const convMean = converted.reduce((sum, val) => sum + val, 0) / converted.length;
    
    const origStd = Math.sqrt(
      original.reduce((sum, val) => sum + Math.pow(val - origMean, 2), 0) / original.length
    );
    const convStd = Math.sqrt(
      converted.reduce((sum, val) => sum + Math.pow(val - convMean, 2), 0) / converted.length
    );
    
    // Quality decreases with larger differences in statistical properties
    const meanDiff = Math.abs(origMean - convMean);
    const stdDiff = Math.abs(origStd - convStd);
    
    // Base quality score with penalties for statistical differences
    let quality = 0.85; // Good base quality for dimensionality reduction
    quality -= meanDiff * 0.1;
    quality -= stdDiff * 0.1;
    
    return Math.max(0, Math.min(1, quality));
  }

  /**
   * Compute skewness of data distribution
   */
  private static computeSkewness(data: number[], mean: number): number {
    const n = data.length;
    if (n < 3) return 0;
    
    const stdDev = Math.sqrt(
      data.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / n
    );
    
    if (stdDev === 0) return 0;
    
    const skewness = data.reduce((sum, val) => sum + Math.pow((val - mean) / stdDev, 3), 0) / n;
    return skewness;
  }

  /**
   * Compute kurtosis of data distribution
   */
  private static computeKurtosis(data: number[], mean: number, stdDev: number): number {
    const n = data.length;
    if (n < 4 || stdDev === 0) return 0;
    
    const kurtosis = data.reduce((sum, val) => sum + Math.pow((val - mean) / stdDev, 4), 0) / n;
    return kurtosis - 3; // Excess kurtosis
  }
}

// Export utility functions
export const convertEncoding = (
  encoding: FaceEncoding, 
  fromFormat: EncodingFormat, 
  toFormat: EncodingFormat
): ConversionResult => {
  if (fromFormat === toFormat) {
    return {
      success: true,
      encoding: {
        encoding,
        format: fromFormat,
        source_system: fromFormat === 'face_recognition' ? 'FRAMS' : 'Face_Reco',
        created_at: new Date().toISOString()
      }
    };
  }

  if (fromFormat === 'mediapipe' && toFormat === 'face_recognition') {
    return FaceEncodingConverter.mediapipeToFaceRecognition(encoding);
  }

  if (fromFormat === 'face_recognition' && toFormat === 'mediapipe') {
    return FaceEncodingConverter.faceRecognitionToMediapipe(encoding);
  }

  return {
    success: false,
    error: `Unsupported conversion: ${fromFormat} to ${toFormat}`
  };
};

export const validateFaceEncoding = (
  encoding: FaceEncoding, 
  format: EncodingFormat
): { valid: boolean; error?: string } => {
  return FaceEncodingConverter.validateEncoding(encoding, format);
};

File: Face_Reco/core/face_encoding_converter.py
"""
Face Encoding Converter Service
Handles bidirectional conversion between face_recognition (128-dim) and MediaPipe (1404-dim) formats
"""

import numpy as np
from typing import List, Dict, Optional, Union, Tuple
from datetime import datetime
import json


class EncodingFormat:
    FACE_RECOGNITION = "face_recognition"
    MEDIAPIPE = "mediapipe"


class FaceEncodingConverter:
    """Converts between different face encoding formats"""
    
    FACE_RECOGNITION_DIM = 128
    MEDIAPIPE_DIM = 1404  # 468 landmarks × 3 coordinates (x,y,z)
    
    @staticmethod
    def mediapipe_to_face_recognition(mediapipe_encoding: List[float]) -> Dict:
        """
        Convert MediaPipe encoding (1404-dim) to face_recognition format (128-dim)
        
        Args:
            mediapipe_encoding: 1404-dimensional MediaPipe face mesh encoding
            
        Returns:
            Dictionary with conversion result
        """
        try:
            # Validate input
            if not isinstance(mediapipe_encoding, (list, np.ndarray)):
                return {
                    'success': False,
                    'error': 'Input must be a list or numpy array'
                }
            
            if len(mediapipe_encoding) != FaceEncodingConverter.MEDIAPIPE_DIM:
                return {
                    'success': False,
                    'error': f'Expected {FaceEncodingConverter.MEDIAPIPE_DIM}-dimensional encoding, '
                           f'got {len(mediapipe_encoding)}'
                }
            
            # Convert to numpy array for easier manipulation
            mediapipe_array = np.array(mediapipe_encoding, dtype=np.float32)
            
            # Extract key facial landmarks from MediaPipe data
            converted_encoding = []
            
            # 1. Extract nose bridge landmarks (first 30 values representing 10 landmarks)
            nose_landmarks = mediapipe_array[:30]
            converted_encoding.extend(FaceEncodingConverter._compute_statistical_features(nose_landmarks, 10))
            
            # 2. Extract eye region landmarks
            left_eye_landmarks = mediapipe_array[99:474]   # Landmarks 33-158
            right_eye_landmarks = mediapipe_array[1086:1461]  # Landmarks 362-487
            converted_encoding.extend(FaceEncodingConverter._compute_statistical_features(left_eye_landmarks, 15))
            converted_encoding.extend(FaceEncodingConverter._compute_statistical_features(right_eye_landmarks, 15))
            
            # 3. Extract mouth region landmarks
            mouth_landmarks = np.concatenate([
                mediapipe_array[183:234],   # Landmarks 61-78
                mediapipe_array[924:975]    # Landmarks 308-324
            ])
            converted_encoding.extend(FaceEncodingConverter._compute_statistical_features(mouth_landmarks, 12))
            
            # 4. Extract jaw/chin landmarks
            jaw_landmarks = np.concatenate([
                mediapipe_array[42:51],     # Landmarks 14-17
                mediapipe_array[597:600]    # Landmarks 199-200
            ])
            converted_encoding.extend(FaceEncodingConverter._compute_statistical_features(jaw_landmarks, 8))
            
            # 5. Extract cheek landmarks
            cheek_landmarks = np.concatenate([
                mediapipe_array[369:429],   # Landmarks 123-143
                mediapipe_array[1056:1116]  # Landmarks 352-372
            ])
            converted_encoding.extend(FaceEncodingConverter._compute_statistical_features(cheek_landmarks, 15))
            
            # 6. Global facial shape features
            converted_encoding.extend(FaceEncodingConverter._compute_global_features(mediapipe_array))
            
            # Pad or truncate to exactly 128 dimensions
            while len(converted_encoding) < FaceEncodingConverter.FACE_RECOGNITION_DIM:
                converted_encoding.append(0.0)
            
            if len(converted_encoding) > FaceEncodingConverter.FACE_RECOGNITION_DIM:
                converted_encoding = converted_encoding[:FaceEncodingConverter.FACE_RECOGNITION_DIM]
            
            # Normalize the encoding
            normalized_encoding = FaceEncodingConverter._normalize_encoding(converted_encoding)
            
            # Calculate quality score
            quality_score = FaceEncodingConverter._calculate_conversion_quality(
                mediapipe_array.tolist(), 
                normalized_encoding
            )
            
            return {
                'success': True,
                'encoding': {
                    'encoding': normalized_encoding,
                    'format': EncodingFormat.FACE_RECOGNITION,
                    'source_system': 'Face_Reco',
                    'created_at': datetime.utcnow().isoformat() + 'Z',
                    'conversion_quality': quality_score,
                    'original_dimensions': len(mediapipe_encoding)
                },
                'quality_score': quality_score
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f'Conversion failed: {str(e)}'
            }
    
    @staticmethod
    def face_recognition_to_mediapipe(face_recognition_encoding: List[float]) -> Dict:
        """
        Convert face_recognition encoding (128-dim) to MediaPipe format (1404-dim)
        Note: This is a lossy conversion with lower accuracy
        
        Args:
            face_recognition_encoding: 128-dimensional face_recognition encoding
            
        Returns:
            Dictionary with conversion result
        """
        try:
            # Validate input
            if not isinstance(face_recognition_encoding, (list, np.ndarray)):
                return {
                    'success': False,
                    'error': 'Input must be a list or numpy array'
                }
            
            if len(face_recognition_encoding) != FaceEncodingConverter.FACE_RECOGNITION_DIM:
                return {
                    'success': False,
                    'error': f'Expected {FaceEncodingConverter.FACE_RECOGNITION_DIM}-dimensional encoding, '
                           f'got {len(face_recognition_encoding)}'
                }
            
            # Convert to numpy array
            face_rec_array = np.array(face_recognition_encoding, dtype=np.float32)
            
            # Initialize MediaPipe encoding array
            mediapipe_encoding = np.zeros(FaceEncodingConverter.MEDIAPIPE_DIM, dtype=np.float32)
            
            # Distribute the 128 features across 468 landmark positions
            features_per_landmark = FaceEncodingConverter.FACE_RECOGNITION_DIM // 468
            remaining_features = FaceEncodingConverter.FACE_RECOGNITION_DIM % 468
            
            feature_index = 0
            
            # Assign features to landmarks
            for landmark in range(468):
                start_index = landmark * 3
                end_index = min(start_index + 3, FaceEncodingConverter.MEDIAPIPE_DIM)
                
                # Assign available features to x,y,z coordinates
                for coord in range(start_index, end_index):
                    if feature_index < FaceEncodingConverter.FACE_RECOGNITION_DIM:
                        mediapipe_encoding[coord] = face_rec_array[feature_index]
                        feature_index += 1
            
            # Distribute remaining features
            for i in range(remaining_features):
                if feature_index < FaceEncodingConverter.FACE_RECOGNITION_DIM:
                    position = int((i / remaining_features) * FaceEncodingConverter.MEDIAPIPE_DIM)
                    mediapipe_encoding[position] = face_rec_array[feature_index]
                    feature_index += 1
            
            # Apply smoothing to make the encoding more realistic
            for i in range(3, len(mediapipe_encoding) - 3):
                mediapipe_encoding[i] = np.mean(mediapipe_encoding[i-3:i+4])
            
            # Normalize to valid coordinate ranges (0-1 for MediaPipe)
            normalized_encoding = FaceEncodingConverter._normalize_to_coordinate_range(
                mediapipe_encoding.tolist()
            )
            
            return {
                'success': True,
                'encoding': {
                    'encoding': normalized_encoding,
                    'format': EncodingFormat.MEDIAPIPE,
                    'source_system': 'FRAMS',
                    'created_at': datetime.utcnow().isoformat() + 'Z',
                    'conversion_quality': 0.65,  # Lower quality due to lossy conversion
                    'original_dimensions': len(face_recognition_encoding)
                },
                'quality_score': 0.65
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f'Conversion failed: {str(e)}'
            }
    
    @staticmethod
    def validate_encoding(encoding: List[float], format_type: str) -> Dict:
        """
        Validate face encoding format and integrity
        
        Args:
            encoding: Face encoding to validate
            format_type: Expected format ('face_recognition' or 'mediapipe')
            
        Returns:
            Validation result dictionary
        """
        if not isinstance(encoding, (list, np.ndarray)):
            return {'valid': False, 'error': 'Encoding must be a list or numpy array'}
        
        if format_type == EncodingFormat.FACE_RECOGNITION:
            if len(encoding) != FaceEncodingConverter.FACE_RECOGNITION_DIM:
                return {
                    'valid': False,
                    'error': f'Face recognition encoding must be {FaceEncodingConverter.FACE_RECOGNITION_DIM}-dimensional'
                }
        elif format_type == EncodingFormat.MEDIAPIPE:
            if len(encoding) != FaceEncodingConverter.MEDIAPIPE_DIM:
                return {
                    'valid': False,
                    'error': f'MediaPipe encoding must be {FaceEncodingConverter.MEDIAPIPE_DIM}-dimensional'
                }
        else:
            return {'valid': False, 'error': 'Invalid format type'}
        
        # Check for NaN or infinite values
        encoding_array = np.array(encoding)
        if not np.isfinite(encoding_array).all():
            return {'valid': False, 'error': 'Encoding contains invalid numeric values'}
        
        return {'valid': True}
    
    @staticmethod
    def _compute_statistical_features(landmarks: np.ndarray, num_features: int) -> List[float]:
        """Compute statistical features from landmark data"""
        if len(landmarks) == 0:
            return [0.0] * num_features
        
        features = []
        
        # Mean
        mean_val = np.mean(landmarks)
        features.append(float(mean_val))
        
        # Standard deviation
        std_val = np.std(landmarks)
        features.append(float(std_val))
        
        # Min and Max
        features.append(float(np.min(landmarks)))
        features.append(float(np.max(landmarks)))
        
        # Range
        features.append(float(np.max(landmarks) - np.min(landmarks)))
        
        # Additional statistical moments
        skewness = FaceEncodingConverter._compute_skewness(landmarks, mean_val)
        features.append(float(skewness))
        
        kurtosis = FaceEncodingConverter._compute_kurtosis(landmarks, mean_val, std_val)
        features.append(float(kurtosis))
        
        # Percentile features
        sorted_landmarks = np.sort(landmarks)
        features.append(float(sorted_landmarks[int(len(sorted_landmarks) * 0.25)]))  # 25th percentile
        features.append(float(sorted_landmarks[int(len(sorted_landmarks) * 0.75)]))  # 75th percentile
        
        # Pad or truncate to required number of features
        while len(features) < num_features:
            features.append(0.0)
        
        return features[:num_features]
    
    @staticmethod
    def _compute_global_features(mediapipe_encoding: np.ndarray) -> List[float]:
        """Compute global facial features"""
        features = []
        
        # Overall facial width (distance between leftmost and rightmost points)
        x_coords = mediapipe_encoding[::3]  # Every 3rd element starting from index 0
        width = float(np.max(x_coords) - np.min(x_coords))
        features.append(width)
        
        # Overall facial height (distance between topmost and bottommost points)
        y_coords = mediapipe_encoding[1::3]  # Every 3rd element starting from index 1
        height = float(np.max(y_coords) - np.min(y_coords))
        features.append(height)
        
        # Facial depth variation
        z_coords = mediapipe_encoding[2::3]  # Every 3rd element starting from index 2
        depth = float(np.max(z_coords) - np.min(z_coords))
        features.append(depth)
        
        return features
    
    @staticmethod
    def _normalize_encoding(encoding: List[float]) -> List[float]:
        """Normalize encoding to standard range"""
        if len(encoding) == 0:
            return encoding
        
        encoding_array = np.array(encoding)
        mean_val = np.mean(encoding_array)
        std_val = np.std(encoding_array)
        
        # Avoid division by zero
        if std_val == 0:
            return [0.0] * len(encoding)
        
        normalized = (encoding_array - mean_val) / std_val
        return normalized.tolist()
    
    @staticmethod
    def _normalize_to_coordinate_range(encoding: List[float]) -> List[float]:
        """Normalize to MediaPipe coordinate range (0-1)"""
        if len(encoding) == 0:
            return encoding
        
        encoding_array = np.array(encoding)
        min_val = np.min(encoding_array)
        max_val = np.max(encoding_array)
        range_val = max_val - min_val
        
        if range_val == 0:
            return [0.5] * len(encoding)
        
        normalized = (encoding_array - min_val) / range_val
        return normalized.tolist()
    
    @staticmethod
    def _calculate_conversion_quality(original: List[float], converted: List[float]) -> float:
        """Calculate conversion quality score"""
        orig_array = np.array(original)
        conv_array = np.array(converted)
        
        orig_mean = np.mean(orig_array)
        conv_mean = np.mean(conv_array)
        
        orig_std = np.std(orig_array)
        conv_std = np.std(conv_array)
        
        # Quality decreases with larger differences in statistical properties
        mean_diff = abs(orig_mean - conv_mean)
        std_diff = abs(orig_std - conv_std)
        
        # Base quality score with penalties for statistical differences
        quality = 0.85  # Good base quality for dimensionality reduction
        quality -= mean_diff * 0.1
        quality -= std_diff * 0.1
        
        return max(0.0, min(1.0, quality))
    
    @staticmethod
    def _compute_skewness(data: np.ndarray, mean_val: float) -> float:
        """Compute skewness of data distribution"""
        n = len(data)
        if n < 3:
            return 0.0
        
        std_val = np.std(data)
        if std_val == 0:
            return 0.0
        
        skewness = np.mean(((data - mean_val) / std_val) ** 3)
        return float(skewness)
    
    @staticmethod
    def _compute_kurtosis(data: np.ndarray, mean_val: float, std_val: float) -> float:
        """Compute kurtosis of data distribution"""
        n = len(data)
        if n < 4 or std_val == 0:
            return 0.0
        
        kurtosis = np.mean(((data - mean_val) / std_val) ** 4)
        return float(kurtosis - 3)  # Excess kurtosis


def convert_encoding(encoding: List[float], from_format: str, to_format: str) -> Dict:
    """
    Convert encoding between formats
    
    Args:
        encoding: Face encoding to convert
        from_format: Source format ('face_recognition' or 'mediapipe')
        to_format: Target format ('face_recognition' or 'mediapipe')
        
    Returns:
        Conversion result dictionary
    """
    if from_format == to_format:
        return {
            'success': True,
            'encoding': {
                'encoding': encoding,
                'format': from_format,
                'source_system': 'FRAMS' if from_format == EncodingFormat.FACE_RECOGNITION else 'Face_Reco',
                'created_at': datetime.utcnow().isoformat() + 'Z'
            }
        }
    
    if from_format == EncodingFormat.MEDIAPIPE and to_format == EncodingFormat.FACE_RECOGNITION:
        return FaceEncodingConverter.mediapipe_to_face_recognition(encoding)
    
    if from_format == EncodingFormat.FACE_RECOGNITION and to_format == EncodingFormat.MEDIAPIPE:
        return FaceEncodingConverter.face_recognition_to_mediapipe(encoding)
    
    return {
        'success': False,
        'error': f'Unsupported conversion: {from_format} to {to_format}'
    }


def validate_face_encoding(encoding: List[float], format_type: str) -> Dict:
    """
    Validate face encoding format and integrity
    
    Args:
        encoding: Face encoding to validate
        format_type: Expected format
        
    Returns:
        Validation result dictionary
    """
    return FaceEncodingConverter.validate_encoding(encoding, format_type)


# Example usage
if __name__ == "__main__":
    # Test conversion from MediaPipe to face_recognition
    test_mediapipe = [0.1] * 1404  # Mock MediaPipe encoding
    result = convert_encoding(test_mediapipe, EncodingFormat.MEDIAPIPE, EncodingFormat.FACE_RECOGNITION)
    print("MediaPipe to face_recognition conversion:", result['success'])
    
    if result['success']:
        print("Converted encoding length:", len(result['encoding']['encoding']))
        print("Quality score:", result['quality_score'])
    
    # Test conversion from face_recognition to MediaPipe
    test_face_rec = [0.2] * 128  # Mock face_recognition encoding
    result = convert_encoding(test_face_rec, EncodingFormat.FACE_RECOGNITION, EncodingFormat.MEDIAPIPE)
    print("\nface_recognition to MediaPipe conversion:", result['success'])
    
    if result['success']:
        print("Converted encoding length:", len(result['encoding']['encoding']))
        print("Quality score:", result['quality_score'])